src/intervals/construction.jl

```diff
@@ -0,0 +1,216 @@
+    lo::T
+    hi::T
+
+    function Interval{T}(a, b) where T
+        a = _normalisezero(a)
+        b = _normalisezero(b)
+        new{T}(T(a, RoundDown), T(b, RoundUp))
+    end
+
+    function Interval{T}(a::T, b::T) where T
+        a = _normalisezero(a)
+        b = _normalisezero(b)
+        new{T}(a, b)
+    end
+end
+
+#= Outer constructors =#
+Interval{T}(a) where T = Interval{T}(a, a)
+Interval(a) = Interval(a, a)
+Interval(a::Tuple) = Interval(a...)
+
+Interval(a::T, b::S) where {T<:AbstractFloat, S} = Interval{promote_type(T, S)}(a, b)
+Interval(a::T, b::S) where {T, S<:AbstractFloat} = Interval{promote_type(T, S)}(a, b)
+Interval(a::T, b::S) where {T<:AbstractFloat, S<:AbstractFloat} = Interval{promote_type(T, S)}(a, b)
+Interval(a::T, b::T) where {T<:AbstractFloat} = Interval{T}(a, b)
+
+Interval(a::T, b::S) where {T, S} = Interval{promote_type(default_bound(), T, S)}(a, b)
+Interval(a::T, b::T) where T = Interval{promote_type(default_bound(), T)}(a, b)
+
+#= Irrational =#
+# Single argument Irrational constructor are in IntervalArithmetic.jl
+# as generated functions need to be define last.
+Interval(a::Irrational, b::T) where {T<:AbstractFloat} = Interval{T}(T(a, RoundDown), b)
+Interval(a::T, b::Irrational) where {T<:AbstractFloat} = Interval{T}(a, T(b, RoundUp))
+
+function Interval(a::Irrational, b::Irrational)
+    T = default_bound()
+    return Interval{T}(a, b)
+end
+
+#= Interval =#
+Interval{T}(x::Interval{T}) where T = x
+Interval{T}(x::Interval) where T = Interval{T}(x.lo, x.hi)
+
+#= Complex =#
+Interval(x::Complex) = Interval(real(x)) + im*Interval(imag(x))
+
+# These definitions has been put there because generated functions must be
+# defined after all methods they use.
+@generated function Interval{T}(x::Irrational) where T
+    res = Interval{T}(x(), x())  # Precompute the interval
+    return :(return $res)  # Set body of the function to return the precomputed result
+end
+
+Interval(x::Irrational) = Interval{default_bound()}(x)
+
+"""
+    interval(a, b)
+
+Create an interval, checking whether [a, b] is a valid `Interval`
+If so, then an `Interval(a, b)` object is returned;
+if not, a warning is printed and the empty interval is returned.
+"""
+function interval(a::T, b::S) where {T<:Real, S<:Real}
+    if !is_valid_interval(a, b)
+        @warn "Invalid input, empty interval is returned"
+        return emptyinterval(promote_type(T, S))
+    end
+
+    return Interval(a, b)
+end
+
+interval(a::Real) = interval(a, a)
+
+const checked_interval = interval
+
+"""
+    a..b
+    ..(a, b)
+
+Create the interval `[a, b]`.
+
+Validity of the interval is checked, but nothing is done to compensate for
+the fact that floating point literals are rounded to nearest when parsed.
+
+Use the string macro `I"[a, b]"` to ensure tight enclosure around the number
+that is typed in, even when it is not exactly representable as a floating point
+number (like `0.1`).
+"""
+..(a, b) = checked_interval(a, b)
```

Comment:
I must admit that I like the current behavior of `..`, i.e., that it creates a directly-rounded interval, which in this PR would be replaced by the string macro.

I simply would like to open that discussion, admitting that the string macro is a good replacement.

Reply:
To open this discussion (in favor of keeping the current behavior for `..`), the string macro is not quite handy if you compute (inside a function) a value stored in `x`, a `Float64`, and you want to *widen it* into the tightest interval that contains it; the function `..` in this case is really handy. Certainly, the macro `@interval(x)` does the job, but at some point (somewhere in the past) I recall the idea was to get rid of rely on it.

Reply:
I think if `interval` and `..` behave the same (as they do after this PR), it would be pointless to keep them both, hence I think either of them can implement the "mind-reading black-magic" that `..` currently does. I don't have a strong opinion on which one should